mlflow/data/numpy_dataset.py

```diff
@@ -0,0 +1,156 @@
+import hashlib
+import json
+from typing import List, Optional, Any, Dict, Union
+
+import numpy as np
+import pandas as pd
+from functools import cached_property
+
+from mlflow.data.dataset import Dataset
+from mlflow.data.dataset_source import DatasetSource
+from mlflow.data.filesystem_dataset_source import FileSystemDatasetSource
+from mlflow.data.pyfunc_dataset_mixin import PyFuncConvertibleDatasetMixin, PyFuncInputsOutputs
+from mlflow.types import Schema
+from mlflow.types.utils import _infer_schema
+
+
+class NumpyDataset(Dataset, PyFuncConvertibleDatasetMixin):
+    """
+    Represents a NumPy dataset for use with MLflow Tracking.
+    """
+
+    def __init__(
+        self,
+        features: Union[np.ndarray, List[np.ndarray], Dict[str, np.ndarray]],
+        source: FileSystemDatasetSource,
+        targets: Union[np.ndarray, List[np.ndarray], Dict[str, np.ndarray]] = None,
+        name: Optional[str] = None,
+        digest: Optional[str] = None,
+    ):
+        """
+        :param features: A numpy array or list/dict of arrays containing dataset features.
+        :param source: The source of the numpy dataset.
+        :param targets: TA numpy array or list/dict of arrays containing dataset targets. Optional
+        :param name: The name of the dataset. E.g. "wiki_train". If unspecified, a name is
+                     automatically generated.
+        :param digest: The digest (hash, fingerprint) of the dataset. If unspecified, a digest
+                       is automatically computed.
+        """
+        self._features = features
+        self._targets = targets
+        super().__init__(source=source, name=name, digest=digest)
+
+    def _compute_digest(self) -> str:
+        """
+        Computes a digest for the dataset. Called if the user doesn't supply
+        a digest when constructing the dataset.
+        """
+        MAX_ROWS = 10000
+
+        flattened_features = self._features.flatten()
+        trimmed_features = flattened_features[0:MAX_ROWS]
+
+        md5 = hashlib.md5()
+
+        # hash trimmed feature contents
+        try:
+            md5.update(pd.util.hash_array(trimmed_features))
+        except TypeError:
+            md5.update(np.int64(trimmed_features.size))
+        # hash full feature dimensions
+        for x in self._features.shape:
+            md5.update(np.int64(x))
+
+        # hash trimmed targets contents
+        if self._targets is not None:
+            flattened_targets = self._targets.flatten()
+            trimmed_targets = flattened_targets[0:MAX_ROWS]
+            try:
+                md5.update(pd.util.hash_array(trimmed_targets))
+            except TypeError:
+                md5.update(np.int64(trimmed_targets.size))
+            # hash full feature dimensions
+            for x in self._targets.shape:
+                md5.update(np.int64(x))
```

Comment:
Nit: Can we DRY this out? The hash logic for targets is the same as for features